File: src/linksearch/classification.py
"""
Structured classification for API consumers: product_type, expected_platforms,
missing_platforms_reason (slug keys). Does not imply every platform will have links.
"""

from __future__ import annotations

# Canonical API slugs (lowercase)
LABEL_TO_SLUG = {
    "Youtube": "youtube",
    "YoutubeShorts": "youtube_shorts",
    "Reddit": "reddit",
    "Tiktok": "tiktok",
    "Instagram": "instagram",
    "Facebook": "facebook",
}

SLUG_TO_LABEL = {v: k for k, v in LABEL_TO_SLUG.items()}

ORDER = ("Youtube", "YoutubeShorts", "Reddit", "Tiktok", "Instagram", "Facebook")


def label_to_slug(label: str) -> str:
    return LABEL_TO_SLUG.get(label, label.lower())


def expected_platform_slugs(
    coverage: dict[str, str],
    requested_labels: set[str],
) -> list[str]:
    """
    Platforms where we predict non-trivial public indexing (high/medium),
    intersected with what the user asked to search.
    """
    out: list[str] = []
    for lab in ORDER:
        if lab not in requested_labels:
            continue
        li = coverage.get(lab, "medium")
        if li in ("high", "medium"):
            out.append(LABEL_TO_SLUG[lab])
    if not out:
        for lab in ORDER:
            if lab in requested_labels:
                out.append(LABEL_TO_SLUG[lab])
    return out


def missing_platform_reasons(
    found_labels: set[str],
    requested_labels: set[str],
    coverage: dict[str, str],
    archetype: str,
    cap_by_label: dict[str, int],
) -> dict[str, str]:
    """Short reasons keyed by slug (e.g. instagram) for requested-but-empty platforms."""
    missing = requested_labels - found_labels
    out: dict[str, str] = {}
    for lab in ORDER:
        if lab not in missing:
            continue
        slug = LABEL_TO_SLUG[lab]
        if cap_by_label.get(lab, 0) <= 0:
            out[slug] = "crawl skipped (low predicted yield for this product)"
            continue
        li = coverage.get(lab, "medium")
        if lab == "Facebook":
            out[slug] = (
                "weak discoverability"
                if li != "low"
                else "low public coverage"
            )
        elif lab in ("Tiktok", "Instagram"):
            out[slug] = "low public coverage" if li == "low" else "weak text match vs thumbnail-heavy posts"
        elif lab == "Reddit":
            out[slug] = "subreddit timing / niche posts" if li != "low" else "low public coverage"
        elif lab in ("Youtube", "YoutubeShorts"):
            out[slug] = "thin descriptions or niche uploads" if li != "low" else "low public coverage"
        else:
            out[slug] = "low public coverage"
        if archetype == "social_heavy" and lab == "Facebook":
            out[slug] = "weak discoverability"
    return out


def build_classification_block(
    found_labels: set[str],
    requested_labels: set[str],
    coverage: dict[str, str],
    archetype: str,
    cap_by_label: dict[str, int],
) -> dict[str, object]:
    return {
        "product_type": archetype,
        "expected_platforms": expected_platform_slugs(coverage, requested_labels),
        "missing_platforms_reason": missing_platform_reasons(
            found_labels, requested_labels, coverage, archetype, cap_by_label
        ),
    }


def social_heavy_zero_tiktok_warn(
    archetype: str,
    requested_labels: set[str],
    found_labels: set[str],
    tiktok_row_count: int,
) -> tuple[bool, str | None]:
    """Reserved hook; user-facing DEBUG TikTok note disabled (no noisy warnings in UI)."""
    return False, None

File: frontend/src/lib/classification.ts
import { effectiveCap, type Budget } from "./coveragePrediction";

const LABEL_TO_SLUG: Record<string, string> = {
  Youtube: "youtube",
  YoutubeShorts: "youtube_shorts",
  Reddit: "reddit",
  Tiktok: "tiktok",
  Instagram: "instagram",
  Facebook: "facebook",
};

const ORDER = [
  "Youtube",
  "YoutubeShorts",
  "Reddit",
  "Tiktok",
  "Instagram",
  "Facebook",
] as const;

export function expectedPlatformSlugs(
  coverage: Record<string, string>,
  requested: Set<string>
): string[] {
  const out: string[] = [];
  for (const lab of ORDER) {
    if (!requested.has(lab)) continue;
    const li = coverage[lab] ?? "medium";
    if (li === "high" || li === "medium") {
      out.push(LABEL_TO_SLUG[lab]);
    }
  }
  if (!out.length) {
    for (const lab of ORDER) {
      if (requested.has(lab)) out.push(LABEL_TO_SLUG[lab]);
    }
  }
  return out;
}

export function missingPlatformReasons(
  found: Set<string>,
  requested: Set<string>,
  coverage: Record<string, string>,
  archetype: string,
  capByLabel: Record<string, number>
): Record<string, string> {
  const missing = [...requested].filter((x) => !found.has(x));
  const out: Record<string, string> = {};
  for (const lab of ORDER) {
    if (!missing.includes(lab)) continue;
    const slug = LABEL_TO_SLUG[lab];
    if ((capByLabel[lab] ?? 0) <= 0) {
      out[slug] = "crawl skipped (low predicted yield for this product)";
      continue;
    }
    const li = coverage[lab] ?? "medium";
    if (lab === "Facebook") {
      out[slug] = li === "low" ? "low public coverage" : "weak discoverability";
    } else if (lab === "Tiktok" || lab === "Instagram") {
      out[slug] =
        li === "low"
          ? "low public coverage"
          : "weak text match vs thumbnail-heavy posts";
    } else if (lab === "Reddit") {
      out[slug] =
        li === "low" ? "low public coverage" : "subreddit timing / niche posts";
    } else {
      out[slug] =
        li === "low" ? "low public coverage" : "thin descriptions or niche uploads";
    }
    if (archetype === "social_heavy" && lab === "Facebook") {
      out[slug] = "weak discoverability";
    }
  }
  return out;
}

export function buildClassification(
  found: Set<string>,
  requested: Set<string>,
  coverage: Record<string, string>,
  archetype: string,
  budgets: Record<string, Budget>,
  baseCap: number
): {
  product_type: string;
  expected_platforms: string[];
  missing_platforms_reason: Record<string, string>;
} {
  const capByLabel: Record<string, number> = {};
  for (const lab of ORDER) {
    const b = budgets[lab];
    if (b) capByLabel[lab] = effectiveCap(baseCap, b);
  }
  return {
    product_type: archetype,
    expected_platforms: expectedPlatformSlugs(coverage, requested),
    missing_platforms_reason: missingPlatformReasons(
      found,
      requested,
      coverage,
      archetype,
      capByLabel
    ),
  };
}

